### PR TITLE
chore(e2e): Port usage of saving favourites including all of the in-use-encryption e2e tests to multiple connections COMPASS-8003

### PR DESCRIPTION
**Pull request by lerouxb**
Main change is that the in-use-encryption tests were ported. browser.shellEval() can now handle multiple commands and properly waits for them to be finished.

### CODE REVIEW

File: packages/compass-e2e-tests/helpers/commands/connect-with-connection-form.ts [hunk 1]
```diff
@@ -1,16 +1,19 @@
...
 export async function connectWithConnectionForm(
   browser: CompassBrowser,
   options: ConnectFormState,
   connectionStatus: 'success' | 'failure' | 'either' = 'success',
   timeout?: number
 ): Promise<void> {
-  const sidebar = await browser.$(Selectors.SidebarTitle);
-  if (await sidebar.isDisplayed()) {
-    await browser.disconnect();
+  await browser.disconnect();
```
Comment (lerouxb) [inline packages/compass-e2e-tests/helpers/commands/connect-with-connection-form.ts:10]: Just made this in line with other places where we disconnect. No need to check anything. We'd have to branch in the checking.

File: packages/compass-e2e-tests/helpers/commands/connect-with-connection-form.ts [hunk 2]
```diff
@@ -1,16 +1,19 @@
...
+
+  // If a connectionName is specified and a connection already exists with this
+  // name, make sure we don't add a duplicate so that tests can always address
+  // this new connection.
+  if (options.connectionName) {
+    await browser.removeConnection(options.connectionName);
```
Comment (lerouxb) [inline packages/compass-e2e-tests/helpers/commands/connect-with-connection-form.ts:16]: Just making sure this mirrors connectWithConnectionString()

File: packages/compass-e2e-tests/helpers/commands/shell-eval.ts [hunk 3]
```diff
@@ -7,10 +7,12 @@ async function getOutputText(browser: CompassBrowser): Promise<string[]> {
   });
 }
 
+let screenshotNumber = 0;
+
 export async function shellEval(
   browser: CompassBrowser,
   connectionName: string,
-  str: string,
+  input: string | string[],
```
Comment (lerouxb) [inline packages/compass-e2e-tests/helpers/commands/shell-eval.ts:13]: shellEval() now supports multiple lines. We used it consecutively in some tests and the problem is that before, we just showed and hid it so it kept its state, but now we open a new one and close it so it loses state. So tests that would `use database` broke because database reset to `test` every time.

Now we can give it multiple lines.

Ideally we'd probably not be using the shell like this for fixture data anyway, but that's probably a larger change.

File: packages/compass-e2e-tests/helpers/commands/shell-eval.ts [hunk 4]
```diff
@@ -19,23 +21,36 @@ export async function shellEval(
...
+        lines.length >
+        /**
+         * input line becomes an output line on enter press, so we are waiting
+         * for two new lines to appear, not just one
+         */
+        numLines + 1
+      ) {
+        // make sure the prompt shows up before entering another line or continueing on
+        await browser
+          .$(`${Selectors.ShellInput} [aria-label="Chevron Right Icon"]`)
```
Comment (lerouxb) [inline packages/compass-e2e-tests/helpers/commands/shell-eval.ts:43]: My proudest achievement 😆  We were typing too fast for some of the commands and this waits for the input to give control back to the user.

File: packages/compass-e2e-tests/tests/connection-form.test.ts [hunk 5]
```diff
@@ -65,8 +65,8 @@ describe('Connection form', function () {
     };
 
     if (TEST_MULTIPLE_CONNECTIONS) {
-      (expectedState.connectionColor = 'no-color'),
-        (expectedState.connectionFavorite = false);
+      expectedState.connectionColor = 'no-color';
```
Comment (lerouxb) [inline packages/compass-e2e-tests/tests/connection-form.test.ts:68]: No idea what prettier did here. I just cleaned it up.

File: packages/compass-e2e-tests/tests/instance-sidebar.test.ts [hunk 6]
```diff
@@ -154,6 +154,7 @@ describe('Instance sidebar', function () {
     }
 
     // open the create database modal from the sidebar
+    // TODO: technically this should be scoped to the connection
```
Comment (lerouxb) [inline packages/compass-e2e-tests/tests/instance-sidebar.test.ts:157]: I'm going go just deal with this in this PR. Hang tight..

File: packages/compass-e2e-tests/tests/in-use-encryption.test.ts [hunk 7]
```diff
@@ -7,26 +7,32 @@ import {
...
 import { MongoClient } from 'mongodb';
 
 import delay from '../helpers/delay';
+import type { ConnectFormState } from '../helpers/connect-form-state';
 
 const CONNECTION_HOSTS = '127.0.0.1:27091';
 const CONNECTION_STRING = `mongodb://${CONNECTION_HOSTS}/`;
 
-async function refresh(browser: CompassBrowser) {
+async function refresh(browser: CompassBrowser, connectionName: string) {
```
Comment (lerouxb) [inline packages/compass-e2e-tests/tests/in-use-encryption.test.ts:22]: This file is where most of the action happens, but github collapses it by default.

File: packages/compass-e2e-tests/tests/instance-sidebar.test.ts [hunk 8]
```diff
@@ -154,9 +154,16 @@ describe('Instance sidebar', function () {
     }
 
     // open the create database modal from the sidebar
-    await browser.clickVisible(Sidebar.CreateDatabaseButton, {
-      screenshot: 'before-can-create-a-database-and-drop-it-click.png',
-    });
+    if (TEST_MULTIPLE_CONNECTIONS) {
```
Comment (lerouxb) [inline packages/compass-e2e-tests/tests/instance-sidebar.test.ts:157]: These are drive-bys: I realised we weren't properly scoping the selector to the relevant connection.

File: packages/compass-e2e-tests/helpers/commands/connect-with-connection-form.ts (hunk 1 above)
Comment (paula-stacho) [reply packages/compass-e2e-tests/helpers/commands/connect-with-connection-form.ts:10]: I wonder if we could name the disconnect action differently then, something like ensureNoActiveConnections. sounds weird, but without your comment and checking the disconnect implementation I'd not think it works like it does.

Comment (lerouxb) [reply packages/compass-e2e-tests/helpers/commands/connect-with-connection-form.ts]: I renamed it to disconnectAll, but I'm generally reserving those kinds of refactoring/distinctions for when we add multiple connections use cases and start differentiating actions like that.

File: packages/compass-e2e-tests/helpers/commands/shell-eval.ts (hunk 4 above)
Comment (paula-stacho) [reply packages/compass-e2e-tests/helpers/commands/shell-eval.ts:43]: nice!

**Review by paula-stacho — APPROVED** [review]
🎉

**Review by addaleax — APPROVED** [review]

File: packages/compass-e2e-tests/helpers/commands/shell-eval.ts [hunk 9]
```diff
@@ -19,23 +19,34 @@ export async function shellEval(
...
+    // wait until more output appears
+    await browser.waitUntil(async () => {
+      const lines = await getOutputText(browser);
+      if (
+        lines.length >
+        /**
+         * input line becomes an output line on enter press, so we are waiting
+         * for two new lines to appear, not just one
+         */
+        numLines + 1
```
[the remark this answers is not in the document]
Comment (lerouxb) [reply packages/compass-e2e-tests/helpers/commands/shell-eval.ts:39]: aah good catch. missed that when adding the loop. probably doesn't matter because the chevron goes away fast enough (in which case the line count check is probably not necessary anymore), but best to move it anyway.
